File: services/mini-coupang/src/main/java/com/deepblog/minicoupang/domain/seller/Seller.java
package com.deepblog.minicoupang.domain.seller;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

/**
 * Seller aggregate root.
 *
 * 소스 패턴(woowahan-eats UserOrder 기반):
 * - @NoArgsConstructor(PROTECTED): JPA 요구사항을 충족하되 외부 new 차단.
 * - @AllArgsConstructor(PRIVATE): @Builder가 내부적으로 사용.
 * - @Builder(toBuilder = true): toBuilder로 불변식 유지하며 파생 인스턴스 생성.
 * - 필드 전체 private + @Getter만 노출. Setter는 금지.
 * - 상태 변경은 반드시 도메인 메서드 경유(rename 등). 직접 필드 접근 X.
 */
@Entity
@Table(name = "sellers")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Builder(toBuilder = true)
public class Seller {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "seller_id")
    private Long id;

    @Column(nullable = false, length = 80)
    private String name;

    @Column(nullable = false, unique = true, length = 200)
    private String email;

    @Version
    private Long version;

    /**
     * 판매자 생성 팩토리. email은 lowercase로 정규화.
     */
    public static Seller create(String name, String email) {
        validateName(name);
        validateEmail(email);
        return Seller.builder()
                .name(name.trim())
                .email(email.trim().toLowerCase())
                .build();
    }

    public void rename(String newName) {
        validateName(newName);
        this.name = newName.trim();
    }

    private static void validateName(String name) {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("name must not be blank");
        }
    }

    private static void validateEmail(String email) {
        if (email == null || email.isBlank()) {
            throw new IllegalArgumentException("email must not be blank");
        }
    }
}

File: services/seller-service/src/main/kotlin/com/deepblog/seller/domain/Seller.kt
package com.deepblog.seller.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import jakarta.persistence.Version

/**
 * Seller 도메인 엔티티.
 *
 * 설계 메모:
 * - `class`(final이 아님) 대신 kotlin-jpa plugin이 open 처리: Hibernate 프록시가 서브클래스로 지연 로딩.
 * - 생성자에 var id = null: JPA 요구사항(기본 생성자)을 kotlin-jpa 플러그인이 자동 제공.
 * - @Version optimistic locking: 여러 요청이 동시에 같은 판매자를 수정할 때 충돌 감지.
 */
@Entity
@Table(name = "sellers")
class Seller(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Column(nullable = false, length = 80)
    var name: String,

    @Column(nullable = false, unique = true, length = 200)
    var email: String,

    @Version
    var version: Long = 0,
)
